scikeras/wrappers.py

```diff
@@ -860,203 +846,61 @@ def scorer(y_true, y_pred, **kwargs) -> float:
-                    # result from a single sigmoid output
-                    # reformat so that we have 2 columns
-                    y[i] = np.column_stack([1 - y[i], y[i]])
-            elif target_type_ in ("multiclass", "multiclass-multioutput"):
-                # array([0.8, 0.1, 0.1], [.1, .8, .1]) ->
-                # array(['apple', 'orange'])
-                idx = np.argmax(y[i], axis=-1)
-                y_ = np.zeros(y[i].shape, dtype=int)
-                y_[np.arange(y[i].shape[0]), idx] = 1
-                if y_.shape[1] == 1:
-                    # Appease the demands of sklearn transformers
-                    y_ = np.squeeze(y_, axis=1)
-                class_predictions.append(self.encoders_[i].inverse_transform(y_))
-            elif target_type_ == "multilabel-indicator":
-                class_predictions.append(
-                    self.encoders_[i].inverse_transform(np.argmax(y[i], axis=1))
-                )
-
-        class_probabilities = np.squeeze(np.column_stack(y))
-
-        y = np.squeeze(np.column_stack(class_predictions))
-
-        # type cast back to input dtype
-        y = y.astype(self.y_dtype_, copy=False)
-
-        extra_args = {"class_probabilities": class_probabilities}
-
-        return y, extra_args
+        return FunctionTransformer()
 
     def _check_output_model_compatibility(self, y):
-        """Checks that the model output number and loss functions match y.
+        """Checks that the model output number and loss functions match
+        what SciKeras expects.
         """
-        # check loss function to adjust the encoding of the input
-        # we need to do this to mimick scikit-learn behavior
-        if isinstance(self.model_.loss, list):
-            losses = self.model_.loss
-        else:
-            losses = [self.model_.loss] * self.n_outputs_
-        for i, loss in enumerate(losses):
-            if is_categorical_crossentropy(loss) and (
-                y[i].ndim == 1 or y[i].shape[1] == 1
-            ):
-                encoder = OneHotEncoder(sparse=False, dtype=np.uint8)
-                tf1dto2d = LabelDimensionTransformer()
-                y[i] = tf1dto2d.fit_transform(y[i])
-                y[i] = encoder.fit_transform(y[i])
-                self.encoders_[i] = make_pipeline(
-                    self.encoders_[i], tf1dto2d, encoder, "passthrough",
-                )
-
-        return super()._check_output_model_compatibility(y)
+        super()._check_output_model_compatibility(y)
+
+        # check that if the user gave us a loss function it ended up in
+        # the actual model
+        if self.loss is not None:
+            try:
+                given = loss_name(self.loss)
+                got = loss_name(self.model_.loss)
+                if got is not given:
+                    warnings.warn(
+                        f"loss={self.loss} but model compiled with {self.model_.loss}."
+                        " Data may not match loss function!"
```

Comment:
It would break the old API, and I think it's a bit too restrictive. If users want to ignore the `loss` constructor parameter and hardcode their loss functions within `build_fn` or do arbitrary logic to select a loss function (i.e. the old API) they should be able to do so without modifying the SciKeras source code. A warning can be filtered or ignored, an error forces you to modify the source code. Thus I am opposed to raising an error here.

Reply:
That's not the case I'm talking about. I'm saying an error should be raised when the user specifies `loss` but that's not reflected in the compiled model. It's this use case:

``` python
>>> wrapper = KerasRegressor(..., loss="mean_abs_error")
>>> wrapper._initialize()
>>> wrapper.model_.loss
"mse"
```

I think an error should be raised in this use case: the user passed a different loss in but it's not reflected in the model.

Reply:
Oh, I see. I am much more open to that, but I feel that it still might lead to other complexities. Would you be okay with leaving that for a future PR?

Reply:
This is what I think the implementation would look like:
```python3
default_val = inspect.signature(self.__init__).parameters["loss"].default
try:
    given = loss_name(self.loss)
    got = loss_name(self.model_.loss)
except ValueError:
    # unknown loss (ex: list of loss functions or custom loss)
    return
if given is not default_val and got is not given:
    raise ValueError
```

We need to check the default value so that if a user does:
```python3
class MyClf(KerasClassifier):
    def __init__(self, loss="different_default", ...)
        super().__init__(self, loss=loss, ...)
```

They can do that without also having to override `_check_output_model_compatibility`.

I think this also applies to a warning, but less so because at least it can be ignored or suppressed elsewhere.

If you're okay with that implementation, I'll put it in. Let me know.

Reply:
Whoops, I filed #100. That'll be closed by this PR then.

Reply:
I guess so!

Reply:
For all the classes in this library, you could hardcode the default to be `loss=None`. The inspection is only relevant if the model loss doesn't match `self.loss`. I'm okay with it because that's a pretty narrow use case.

``` python
if loss_name(self.loss) != loss_name(self.model_.loss):
    warn(...)
    default_loss = inspect.signature(self.__init__).parameters["loss"].default
    if default_loss != loss_name(self.loss):
        raise ValueError(...)
```

Reply:
How would you feel about having this check only for non-subclassed models? Then it would be:

```python3
if self.__class__ is KerasClassifier:
    if self.loss is not None and loss_name(self.loss) != loss_name(self.model_.loss):
        ValueError(...)
```

Reply:
:-1: this class shouldn't raise any errors:

``` python
class NoErrors(KerasClassifier):
    pass
```

Reply:
I don't think the implementation in https://github.com/adriangb/scikeras/pull/88#discussion_r502822990 would raise an error for the example in https://github.com/adriangb/scikeras/pull/88#discussion_r502842545, but I could be wrong. But in any case, I did not implement https://github.com/adriangb/scikeras/pull/88#discussion_r502822990, so are we okay with the current implementation then?

Reply:
Since https://github.com/adriangb/scikeras/pull/88#discussion_r502822990 was never implemented and the previous implementation seemed to be uncontroversial, I'm resolving this.